### PR TITLE
tests/main/buildmode: verify usability of PIE hardening for deb packages

### PR DESCRIPTION
**Pull request by bboozzoo**
Enable PIE hardening flags when building snapd package.

This currently fails because we attempt link a static variant of libapparmor, but the library was not built with -fPIC apparently.  See [LP#1824384](https://bugs.launchpad.net/ubuntu/+source/apparmor/+bug/1824384).

The relevant build log:
```
mv -f snap-confine/.deps/snap_confine_snap_confine-user-support.Tpo snap-confine/.deps/snap_confine_snap_confine-user-support.Po                                             
gcc -Wall -Wextra -Wmissing-prototypes -Wstrict-prototypes -Wno-missing-field-initializers -Wno-unused-parameter -Werror  -DLIBEXECDIR=\"/usr/lib/snapd\" -DNATIVE_LIBDIR=\"/usr/lib\"     -g -O2 -fPIE -fstack-protector-strong -Wformat -Werror=format-security   -Wl,-Bsymbolic-functions -fPIE -pie -Wl,-z,relro -o snap-confine/snap-confine snap-confine/snap_confine_snap_confine-cookie-support.o snap-confine/snap_confine_snap_confine-mount-support-nvidia.o snap-confine/snap_confine_snap_confine-mount-support.o snap-confine/snap_confine_snap_confine-ns-support.o snap-confine/snap_confine_snap_confine-seccomp-support-ext.o snap-confine/snap_confine_snap_confine-seccomp-support.o snap-confine/snap_confine_snap_confine-snap-confine-args.o snap-confine/snap_confine_snap_confine-snap-confine-invocation.o snap-confine/snap_confine_snap_confine-snap-confine.o snap-confine/snap_confine_snap_confine-udev-support.o snap-confine/snap_confine_snap_confine-user-support.o  libsnap-confine-private.a -ludev    -Wl,-Bstatic -lcap -lapparmor  -Wl,-Bdynamic -pthread
/usr/bin/ld: /lib/x86_64-linux-gnu/libapparmor.a(kernel.o): 
  relocation R_X86_64_32 against `.rodata.str1.1'
  can not be used when making a shared object; recompile with -fPIC 
/lib/x86_64-linux-gnu/libapparmor.a: error adding symbols: Bad value
collect2: error: ld returned 1 exit status
```

### CODE REVIEW

**Comment by stolowski** [comment]
What's the plan for this PR?

**Comment by bboozzoo** [comment]
File a bug to apparmor. Perhaps @jdstrand know more about whether there's a chance to see it addressed in near future.

**Comment by jdstrand** [comment]
> File a bug to apparmor. Perhaps @jdstrand know more about whether there's a chance to see it addressed in near future.

I asked upstream and will respond in the bug when I hear back.

**Comment by pedronis** [comment]
@bboozzoo could we write a quick spread test that will fail once the the static library is in the mode we want? then the test can contain a link to this PR that can be closed. When/ifs that test fails we can reopen this.

**Comment by bboozzoo** [comment]
Closing for now. Will reopen with a spread test that builds snapd.

File: tests/main/buildmode-pie/task.yaml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        objdump -f $DIRTY_SOURCE/snapd/debian/snapd/usr/lib/snapd/snap-confine | MATCH 'DYNAMIC'
+        # make sure that the we did not expect the current system to fail
+        for n in $FAILING_SYSTEMS; do
+           if echo "$SPREAD_SYSTEM" | grep -q "$n" ; then
+               echo "unexpected success on system $SPREAD_SYSTEM"
+               exit 1
+           fi
+        done
+    else
+        # make sure that the we expected the current system to fail
```
Comment (zyga) [inline tests/main/buildmode-pie/task.yaml]:
```suggestion
        # Make sure that we expected the current system to fail.
```

File: tests/main/buildmode-pie/task.yaml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    cat <<'EOF' > debian/testflags.mk
+    export DEB_BUILD_MAINT_OPTIONS = hardening=+pie
+    include /usr/share/dpkg/buildflags.mk
+    EOF
+    sed -e 's|# include /usr/share/dpkg/buildflags.mk|include debian/testflags.mk|' -i debian/rules
+    chown -R test:test "$DIRTY_SOURCE/snapd"
+    if su -l -c "cd $DIRTY_SOURCE/snapd && DEB_BUILD_OPTIONS='nocheck testkeys' dpkg-buildpackage -tc -b -Zgzip" test; then
+        # verify that snap-confine is really PIE
+        objdump -f $DIRTY_SOURCE/snapd/debian/snapd/usr/lib/snapd/snap-confine | MATCH 'DYNAMIC'
+        # make sure that the we did not expect the current system to fail
```
Comment (zyga) [inline tests/main/buildmode-pie/task.yaml]:
```suggestion
        # Make sure that the we did not expect the current system to fail.
```

File: tests/main/buildmode/task.yaml [hunk 3]
```diff
@@ -20,3 +20,7 @@ execute: |
     for p in snapd snap-repair; do
         objdump -f /usr/lib/snapd/"$p" | MATCH "$needle"
     done
+
+    # C binaries are always built in non-PIE mode because we do not enable
+    # hardening
+    objdump -f /usr/lib/snapd/snap-confine | not MATCH 'DYNAMIC'
```
Comment (zyga) [inline tests/main/buildmode/task.yaml]: Can you please add a comment that explains how absence of DYNAMIC is related to PIE.

File: tests/main/buildmode-pie/task.yaml [hunk 4]
```diff
@@ -0,0 +1,47 @@
+summary: Test that snapd deb package can be built using pie hardening flags
+
+systems: [ubuntu-1*, ubuntu-2*, debian-*]
+
+environment:
+    FAILING_SYSTEMS: ubuntu-16.04- ubuntu-18.04- debian-
+    DIRTY_SOURCE: /tmp/dirty-snapd
+
+prepare:
```
Comment (zyga) [inline tests/main/buildmode-pie/task.yaml]:
```suggestion
prepare: |
```

Comment (bboozzoo) [reply tests/main/buildmode-pie/task.yaml]: Thanks!

**Comment by pedronis** [comment]
@bboozzoo  this is actually failing like this:

```
2020-02-18 10:25:13 Error executing google:ubuntu-19.10-64:tests/main/buildmode : 
-----
+ needle=DYNAMIC
++ getconf LONG_BIT
+ '[' 64 = 32 ']'
+ MATCH DYNAMIC
+ objdump -f /usr/bin/snap
+ for p in snapd snap-repair
+ MATCH DYNAMIC
+ objdump -f /usr/lib/snapd/snapd
+ for p in snapd snap-repair
+ MATCH DYNAMIC
+ objdump -f /usr/lib/snapd/snap-repair
+ not MATCH DYNAMIC
+ objdump -f /usr/lib/snapd/snap-confine
```

**Comment by bboozzoo** [comment]
Looks like the behavior is inconsistent and changed between 16.04 and 20.04. 

The following check fails consistently on all the 
```
  eval "$(DEB_BUILD_OPTIONS=hardening=+pie dpkg-buildflags --export=sh)"
  gcc aa.c -static $CPPFLAGS $CFLAGS $LDFLAGS -lapparmor -pthread
  objdump -f ./a.out | MATCH DYNAMIC
  hardening-check ./a.out | MATCH 'Position Independent Executable.*: yes'
```

In between there's a `hardening-wrapper` that is no longer present in 18.04 (?).  Anyways, it does seem a problem to be able to build a static binary with PIE enabled and gcc < 8.2. Closing this, until I have more ideas on how to approach the problem.

**Pull request closed without merging**